### PR TITLE
Bump ot with in process propag

### PR DESCRIPTION
**Pull request by pavolloffay**
- [x] Renamed decorators to include Okhttp in the name
- [x] Use BaseSpan in decorators

### CODE REVIEW

**Comment by pavolloffay** [comment]
@jpkrohling could you please review when you have some time? repo includes opentracing aware executor which will be in the separate repository soon.

**Comment by pavolloffay** [comment]
Do not merge yet, this does not work well for multiple async requests.

**Comment by pavolloffay** [comment]
@jpkrohling could you please review? It should be ready :)

**Comment by objectiser** [comment]
@pavolloffay Looking to update the java-agent rules for okhttp, and found that the approach used to add the interceptor has changed in this PR. Any thoughts on how best to update the rules https://github.com/opentracing-contrib/java-agent/blob/master/rules/java-okhttp/src/main/resources/otarules.btm using the new approach?

**Comment by objectiser** [comment]
@pavolloffay Changed rule to use the renamed span decorator and now works - although it creates 2 spans, where previously there was only 1. The additional span is a component that acts as the parent for the client span.

Rather than create the separate component span, would it just be possible to check for active span, and if found then create a continuation - but if no active span, then leave it so the client span becomes the root?

**Comment by pavolloffay** [comment]
> Rather than create the separate component span, would it just be possible to check for active span, and if found then create a continuation - but if no active span, then leave it so the client span becomes the root?

I don't know if I understand it correctly. Could you please describe/model it for a simple case and for redirects? 

This instrumentation creates one local span for the whole invocation and then a client span for each network attempt (with all HTTP and peer tags). Before it was adding logs for each network span (e.g. redirect). I think the current approach is cleaner. e.g. it allows you to query based on tags for each network attempt.

**Comment by objectiser** [comment]
@pavolloffay Ok thanks for the explanation - yes agree better to have separate span for each attempt.

**Comment by pavolloffay** [comment]
Thanks, I see where you are pointing with "one span", however I don't know how to solve it properly for all cases, this seems better than adding logs due to queries and following analysis.

Is it ready to go in from your view?

File: README.md [hunk 1]
```diff
@@ -1,25 +1,19 @@
...
-// create traced request 
-client.newCall(new Request.Builder()
-        .url("http://localhost/foo")
-        // optional: client span started in inteceptor will be child of provided SpanContext
-        // by default client span is in new trace
-        .tag(new TagWrapper(parentSpan.context())) 
-        .build())
-    .execute();
+Call.Factory callFactory = new TracingCallFactory(okHttpClient, tracer);
+client.newCall(request)...
```
Comment (objectiser) [inline README.md]: Should this `client` and the previous `okHttpClient` be the same object? Is the `callFactory` used for anything else?

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingCallFactory.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+             * In case of exception network interceptor is not called
+             */
+            OkHttpClient.Builder okBuilder = okHttpClient.newBuilder();
+            okBuilder.networkInterceptors().add(0, new NetworkInterceptor(tracer, activeSpan.context(), decorators));
+
+            final ActiveSpan.Continuation continuation = activeSpan.capture();
+            okBuilder.interceptors().add(0, new Interceptor() {
+                @Override
+                public Response intercept(Chain chain) throws IOException {
+                    ActiveSpan activate = continuation.activate();
```
Comment (objectiser) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingCallFactory.java]: nit: not sure if `activate` is an appropriate name for this active span. Maybe `activatedSpan`?

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingCallFactory.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                }
+
+                Request.Builder requestBuilder = chain.request().newBuilder();
+                tracer.inject(networkSpan.context(), Format.Builtin.HTTP_HEADERS, new RequestBuilderInjectAdapter(requestBuilder));
+                Response response = chain.proceed(requestBuilder.build());
+
+                for (OkHttpClientSpanDecorator spanDecorator: decorators) {
+                    spanDecorator.onResponse(chain.connection(), response, networkSpan);
+                }
+
```
Comment (objectiser) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingCallFactory.java:119]: Would it not be better to handle the exception (and onError decoration) here, so all decoration handled in the same interceptor?

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java [hunk 4]
```diff
@@ -2,39 +2,55 @@
...
 import okhttp3.Interceptor;
 import okhttp3.OkHttpClient;
 import okhttp3.Request;
 import okhttp3.Response;
 
 /**
- * Okhttp interceptor to trace client requests. Interceptor adds span context into outgoing requests.
- * By default span operation name is set to HTTP method.
+ * OkHttp interceptor to trace client requests. Interceptor adds span context into outgoing requests.
+ * Please this instrumentation only when {@link TracingCallFactory} is not possible to use. This
```
Comment (objectiser) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java]: "Please only use this instrumentation when .....".

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java [hunk 5]
```diff
@@ -2,39 +2,55 @@
...
+ * instrumentation fails properly infer parent span when doing simultaneously asynchronous calls.
```
Comment (objectiser) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java]: "fails to properly"

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java [hunk 6]
```diff
@@ -2,39 +2,55 @@
...
  */
 public class TracingInterceptor implements Interceptor {
-
     private static final Logger log = Logger.getLogger(TracingInterceptor.class.getName());
 
     private Tracer tracer;
-    private List<SpanDecorator> decorators;
+    private List<OkHttpClientSpanDecorator> decorators;
+
+    public static OkHttpClient addTracing(OkHttpClient.Builder builder, Tracer tracer) {
```
Comment (objectiser) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java]: nit: move method after the constructor - prefer to see constructors up front.

**Review by objectiser — APPROVED** [review]
Some minor changes - one thing, would it be better to move setting the `component` tag into the STANDARD_TAGS, as I don't think it is being set on some of the spans.

File: README.md (hunk 1 above)
Comment (pavolloffay) [reply README.md]: https://square.github.io/okhttp/3.x/okhttp/okhttp3/Call.Factory.html it is the main interface to create new calls. OkHttpClient implements it.

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingCallFactory.java (hunk 3 above)
Comment (pavolloffay) [reply opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingCallFactory.java:119]: The problem is that it is not possible to catch that exception here :(

File: README.md (hunk 1 above)
Comment (objectiser) [reply README.md]: So shouldn't it be `callFactory.newCall(request)`?

Comment (pavolloffay) [reply README.md]: +1, I wanted to name it client :+1:

**Comment by pavolloffay** [comment]
@objectiser thanks

**Comment by objectiser** [comment]
@pavolloffay LGTM - although still issue that `component` is not associated with all spans. Realised that the local span does not use the decorator, so having in standard tags won't address that - but atleast if added to standard tags (onRequest) means it would deal with all of the client spans.

**Comment by objectiser** [comment]
Once merged, will you be able to do a release?

**Pull request merged**